doc/rtd/reference/merging.rst

```diff
@@ -3,114 +3,179 @@
-(and not append). So, it was decided to create a new and improved way to merge
-dictionaries (and their contained objects) together in a customisable way,
-thus allowing users who provide cloud-config user data to determine exactly
-how their objects will be merged.
+Merging is not enabled by default. There are two ways to do so:
 
-For example:
 
-.. code-block:: yaml
+1. In cloud-config YAML configuration, two keys will be checked: ``merge_how``,
+   and ``merge_type`` (in this order). Their value should be either a format
+   string, or a configuration object as described below.
+2. In multi-part documents, two headers will be checked: ``Merge-Type`` and
+   ``X-Merge-Type`` (in this order). The value should contain a merging format
+   string as described below.
 
-   #cloud-config (1)
-   runcmd:
-     - bash1
-     - bash2
+There are three types of configuration that can be merged: Lists, dictionaries,
+and strings. Cloud-config allows enabling and configuring merging for all of
+them separately. For the available options, look below.
 
-   #cloud-config (2)
-   runcmd:
-     - bash3
-     - bash4
+Dictionary format
+-----------------
 
-The previous way of merging the two objects above would result in a final
-cloud-config object that contains the following:
+A dictionary can be used when it specifies the same information as the
+string format (i.e., the second option above). For example:
 
 .. code-block:: yaml
 
-   #cloud-config (merged)
-   runcmd:
-     - bash3
-     - bash4
+   merge_how:
+     - name: list
+       settings: ['append']
+     - name: dict
+       settings: ['no_replace', 'recurse_list']
+     - name: str
+       settings: ['append']
 
-Typically this is not what users want - instead they would prefer:
+This example matches the default configuration.
 
-.. code-block:: yaml
+String format
+-------------
 
-   #cloud-config (merged)
-   runcmd:
-     - bash1
-     - bash2
-     - bash3
-     - bash4
+The following string format is expected: ::
 
-This change makes it easier to combine the various cloud-config objects you
-have into a more useful list. In this way, we reduce the duplication necessary
-to accomplish the same result with the previous method.
+   type1(option1,option2)+type2(option3,option4)....
+
+The type must be one of ``list`` (arrays), ``dict`` (objects), 
+or ``str`` (strings). In brackets, you can pass options for each type. To use
+the defaults, pass the type with braces (e.g. ``list()``). The default format
+string looks like this:
+
+.. code-block:: python
+
+   list()+dict()+str()
 
-Built-in mergers
-================
+To enable an option, pass its name (or omit to disable):
 
-``Cloud-init`` provides merging for the following built-in types:
+.. code-block:: python
+
+   list(replace,append)+dict(replace)
 
-- :command:`Dict`
-- :command:`List`
-- :command:`String`
+Available options
+-----------------
 
-``Dict``
---------
+Objects
```

Comment:
I agree that using both is confusing, but wouldn't it make more sense from a user perspective to adopt the yaml convention in the documentation (which would probably include adding aliases to the rule syntax)? I'll update that section anyway, but just wanted to point this out.

Reply:
Changed that back to dict.

Reply:
I hear you, and in theory, I agree, but since there's so much other terminology that maps to python language that would be hard to change without extensive code changes (e.g., `dict(recurse_list)+list(append)+str()`), I still think it's most cohesive to define things in terms of the python terminology. Thanks for changing it back.